File: go/relation.go
/* Copyright (c) 2025 Richard Rodger, MIT License */

package aontu

import (
	"sort"
	"strconv"
	"strings"
)

// RELATION GRAPH VERDICTS (RELATIONS.0.md §3.3, replacing the G4
// phase 5 magic-key pass; the Go side of ts/src/relation.ts):
// acyclicity and inverse consistency over the edge set, DECLARED by
// the graph atoms -- `acyclic()` and `inverse(name)` conjoined at the
// field whose key is the predicate -- and decided AFTER unification,
// never by it.
//
// Why not in the lattice. Both properties are GLOBAL and NON-MONOTONE:
// an acyclic graph becomes cyclic when one more edge unifies in, and an
// inverse that is present becomes absent when the far side is narrowed.
// The lattice guarantee is that more information never falsifies what
// has been observed, so a constraint that could be true and then false
// is not a constraint the lattice may hold. The atoms therefore only
// REGISTER during unification (GraphAtomVal.register, onto
// Ctx.reldecls), and the verdict lands at GENERATION -- the sizing
// atoms' model -- where no more information can arrive. The
// `relations` verb reports the same verdict from the same
// declarations: one decision, two surfaces. The old `relations:`
// magic key is GONE, discharging ADR-010's grandfather clause.

// RelationFinding is one broken relation property.
// Field order is LEXICOGRAPHIC, the canonical emitter's order — the
// TypeScript port's exactJSON sorts keys, and a report is read by a
// machine that diffs it.
type RelationFinding struct {
	// At is where the offending edge is written, as a `$.dotted.path`.
	At   string `json:"at"`
	Code string `json:"code"`
	// Detail is, for a cycle, the entities it runs through in the order
	// the walk found them, closing back on the first; for a missing
	// inverse, the two ends and the relation that should have mirrored
	// it.
	Detail []string `json:"detail"`
	// Relation the finding is about.
	Relation string `json:"relation"`
}

// RelationReport is the relation checks for one document.
type RelationReport struct {
	// Declared is how many relation declarations the document makes
	// (G11 phase 4). ZERO is the answer that matters: `pass` over no
	// declarations means nothing was checked, and without this the two
	// are one word. A POINTER, because `omitempty` would drop the zero
	// that carries the meaning; nil is "the run did not ask", &0 is
	// "asked, and there were none". Absent unless RelationOptions.Count
	// asked for it, so no existing caller's report changes shape.
	Declared *int `json:"declared,omitempty"`
	// Errors is WHY the graph could not be looked at, in the same
	// finding shape Vet reports in (the review's finding F). Findings is
	// about the GRAPH and stays that way; a document that does not stand
	// up has no graph to have findings about, and an `error` verdict
	// used to arrive with an empty list -- something is wrong, and
	// nothing about what. Present ONLY on an `error` verdict.
	Errors   []VetFinding      `json:"errors,omitempty"`
	Findings []RelationFinding `json:"findings"`
	Verdict  string            `json:"verdict"`
}

// RelationOptions are the run's knobs. A nil value is the default run.
type RelationOptions struct {
	// Count fills RelationReport.Declared, so a caller can tell a graph
	// that CHECKED CLEAN from one there was nothing to check. Mirrors
	// the canonical port's RelationOptions.count.
	Count bool
}

// declaredRelation is one declared relation, as the document spells it.
func findCycle(start string, succ map[string][]string, done map[string]bool) []string {
	stack := []string{}
	onStack := map[string]bool{}

	var walk func(node string) []string
	walk = func(node string) []string {
		if onStack[node] {
			for i, n := range stack {
				if n == node {
					return append(append([]string{}, stack[i:]...), node)
				}
			}
		}
		if done[node] {
			return nil
		}
		done[node] = true
		stack = append(stack, node)
		onStack[node] = true
		for _, next := range succ[node] {
			if found := walk(next); nil != found {
				return found
			}
		}
		stack = stack[:len(stack)-1]
		delete(onStack, node)
		return nil
	}

	return walk(start)
}

// RelationCheck runs the relation checks over one document.
// relationFindings is the verdict itself, pure over what the
// evaluation produced: the registered declarations and the edge set.
// Shared by the generation hook (relationErrors) and the `relations`
// verb, so the two surfaces cannot disagree. Mirrors relationFindings
// in ts/src/relation.ts.
func relationFindings(decls map[string]*relDecl, graph Graph) []RelationFinding {
	findings := []RelationFinding{}

	// The edge set, indexed the two ways the checks read it.
	byRelation := map[string][]Edge{}
	pairs := map[string]bool{}
	for _, e := range graph.Edges {
		byRelation[e.Key] = append(byRelation[e.Key], e)
		pairs[e.Key+" "+e.From+" "+e.To] = true
	}

	// Predicates in sorted order, so the findings arrive the same way
	// in both ports (the registry is a Go map, randomly ordered).
	names := make([]string, 0, len(decls))
	for name := range decls {
		names = append(names, name)
	}
	sort.Strings(names)

	for _, name := range names {
		decl := decls[name]
		mine := byRelation[name]

		if decl.acyclic {
			succ := map[string][]string{}
			for _, e := range mine {
				succ[e.From] = append(succ[e.From], e.To)
			}
			roots := make([]string, 0, len(succ))
			for from, list := range succ {
				sort.Strings(list)
				roots = append(roots, from)
			}
			sort.Strings(roots)

			// The roots are visited in sorted order, and a node
			// already settled is not revisited, so one cycle is
			// reported once and the SAME one in both ports.
			done := map[string]bool{}
			for _, from := range roots {
				cycle := findCycle(from, succ, done)
				if nil != cycle {
					// The cycle's first node is a key of succ, and
					// every key of succ came from an edge's From, so
					// the edge is there.
					at := ""
					for _, e := range mine {
						if e.From == cycle[0] {
							at = e.At
							break
						}
					}
					findings = append(findings, RelationFinding{
						Code:     "relation_cycle",
						Relation: name,
						At:       at,
						Detail:   cycle,
					})
					break
				}
			}
		}

		inverses := make([]string, 0, len(decl.inverses))
		for inv := range decl.inverses {
			inverses = append(inverses, inv)
		}
		sort.Strings(inverses)
		for _, inv := range inverses {
			for _, e := range mine {
				to := e.To
				if !pairs[inv+" "+to+" "+e.From] {
					findings = append(findings, RelationFinding{
						Code:     "relation_inverse_missing",
						Relation: name,
						At:       e.At,
						Detail:   []string{e.From, to, inv},
					})
				}
			}
		}
	}

	// SORTED, because a report is read by a machine that diffs it: by
	// the position of the offending edge, then by code, then by the
	// detail (two inverse declarations on one predicate can flag one
	// edge twice).
	sort.SliceStable(findings, func(i, j int) bool {
		a, b := findings[i], findings[j]
		if a.At != b.At {
			return a.At < b.At
		}
		if a.Code != b.Code {
			return a.Code < b.Code
		}
		return strings.Join(a.Detail, " ") < strings.Join(b.Detail, " ")
	})

	return findings
}

// relationErrors is the generation hook (Aontu.GenerateVars, between
// unification success and value generation): the first finding becomes
// a LOCATED evaluation error at the offending edge, exactly as an
// unmet sizing atom refuses at generation. Findings name entities and
// positions the document spelled, so the walk to the site cannot miss.
// Mirrors relationErrors in ts/src/relation.ts (which files every
// finding; the Go generation path returns its first error, as the bag
// walks do).
func relationErrors(ctx *Ctx, root Val) error {
	if nil == ctx || 0 == len(ctx.reldecls) {
		return nil
	}
	findings := relationFindings(ctx.reldecls, GraphOf(root))
	if 0 == len(findings) {
		return nil
	}
	// EVERY finding becomes a collected error, exactly as the TS hook
	// adderr's each one (two inverse declarations on one predicate
	// refuse twice, and the report says both names); makeNilErrFull
	// records on ctx, and errmsg renders them all, `------`-joined,
	// as any multi-error document renders.
	for _, f := range findings {
		node := root
		if 2 < len(f.At) {
			for _, seg := range strings.Split(f.At[2:], ".") {
				// Graph atoms hold the field's value -- possibly nested,
				// one atom carrying another -- and the path steps through
				// them exactly as the graph walk does.
				for {
					ga, ok := node.(*GraphAtomVal)
					if !ok {
						break
					}
					node = ga.held
				}
				switch n := node.(type) {
				case *MapVal:
					node = n.peg[seg]
				case *ListVal:
					ix, _ := strconv.Atoi(seg)
					node = n.peg[ix]
				}
			}
			// No unwrap AFTER the walk: a finding's At names an edge
			// element, and an edge's element is a string -- an
			// atom-wrapped element mints no edge in the first place
			// (the graph visit descends atoms only at field values),
			// so the walk cannot end on an atom.
		}
		makeNilErrFull(ctx, f.Code, node, nil, "relate", map[string]string{
			"relation": f.Relation,
			"detail":   strings.Join(f.Detail, " -> "),
		})
	}
	return &AontuError{Msg: ctx.errmsg(), Code: findings[0].Code}
}

// RelationCheck runs the relation checks over one document: evaluate,
// then report the same verdict generation enforces.
func (a *Aontu) RelationCheck(src string) RelationReport {
	return a.RelationCheckOpts(src, nil)
}

// RelationCheckOpts is RelationCheck with the run's knobs. Split from
// it rather than folded into it because every existing caller wants
// the default run, and the Vet precedent is a package surface whose
// options are a pointer that may be nil.
func (a *Aontu) RelationCheckOpts(
	src string, opts *RelationOptions) RelationReport {
	options := RelationOptions{}
	if nil != opts {
		options = *opts
	}
	// Parsed and unified in two steps rather than through Unify, so the
	// failure can be REPORTED: the context carries the engine's own
	// first error, and Unify hands back only that something went wrong.
	parsed, perr := a.parseEntry(src)
	if nil != perr {
		return RelationReport{Verdict: "error", Findings: []RelationFinding{},
			Errors: []VetFinding{parseFinding(a.File, VetRoleData, perr)}}
	}

	root, ctx, _ := a.unifyCtx(parsed, nil, src)

	// A document that does not stand up is not a document with a bad
	// graph: the errors it already has are the answer, and blaming its
	// relations on top would be noise.
	if nil == root || root.Nil() || 0 < len(ctx.err) {
		return RelationReport{Verdict: "error", Findings: []RelationFinding{},
			Errors: []VetFinding{failureFinding(ctx, a.File, src, root)}}
	}

	// The count rides every report from here down: the engine knows it
	// at exactly this point, and a caller asking "was there anything to
	// check?" should not have to evaluate the document a second time to
	// find out.
	var declared *int
	if options.Count {
		n := len(ctx.reldecls)
		declared = &n
	}
	if 0 == len(ctx.reldecls) {
		return RelationReport{Verdict: "pass", Declared: declared,
			Findings: []RelationFinding{}}
	}

	// NOR IS A DOCUMENT THAT CANNOT BE GENERATED. Unification can
	// succeed over a tree that still holds an unsettled disjunction --
	// more than one alternative admitted, which disjunct_no_gen
	// refuses at generation -- and the graph walk cannot read inside
	// one: the links in every alternative are invisible, so a mirror
	// the document plainly writes is missing from the edge set and
	// inverse(n) calls it missing. Generation is therefore asked
	// FIRST, on a collecting context of its own, and what it says is
	// the answer. Mirrors ts/src/relation.ts.
	gcopy := *ctx
	gctx := &gcopy
	gctx.err = nil
	gctx.collect = true
	root.Gen(gctx)
	if 0 < len(gctx.err) {
		return RelationReport{Verdict: "error", Findings: []RelationFinding{},
			Errors: []VetFinding{failureFinding(gctx, a.File, src, root)}}
	}

	findings := relationFindings(ctx.reldecls, GraphOf(root))
	verdict := "pass"
	if 0 < len(findings) {
		verdict = "fail"
	}
	return RelationReport{Verdict: verdict, Declared: declared,
		Findings: findings}
}

File: ts/src/relation.ts
/* Copyright (c) 2025 Richard Rodger, MIT License */
import { includeOpts } from './utility'


import { Aontu } from './aontu'
import { failureFinding } from './vet'
import type { VetFinding } from './vet'
import type { TrustOptions } from './type'
import { graphOf } from './graph'
import type { Edge, Graph } from './graph'
import { cmpCodePoint } from './keyorder'
import { makeNilErr } from './err'
import type { RelDecl } from './val/GraphAtomVal'


export type RelationVerdict = 'pass' | 'fail' | 'error'

export type RelationFinding = {
  code: string
  relation: string
  // Where the offending edge is written, as a `$.dotted.path`.
  at: string
  detail: string[]
}

export type RelationReport = {
  verdict: RelationVerdict
  findings: RelationFinding[]

  errors?: VetFinding[]

  declared?: number
}

export type RelationOptions = {
  // Where the document CAME FROM, so a relative `@"file"` load inside
  // it resolves from its own directory (trimCheck's precedent).
  path?: string
  trust?: TrustOptions

  // Extensions additionally read as text (the CLI's `--text-ext`).
  // Rides beside `trust` because it is the other half of what an
  // include may read.
  textExt?: string[]

  count?: boolean
}


function findCycle(
  start: string,
  succ: Map<string, string[]>,
  done: Set<string>,
): string[] | undefined {
  const stack: string[] = []
  const onStack = new Set<string>()

  const walk = (node: string): string[] | undefined => {
    if (onStack.has(node)) {
      return [...stack.slice(stack.indexOf(node)), node]
    }
    if (done.has(node)) {
      return undefined
    }
    done.add(node)
    stack.push(node)
    onStack.add(node)
    for (const next of succ.get(node) ?? []) {
      const found = walk(next)
      if (undefined !== found) {
        return found
      }
    }
    stack.pop()
    onStack.delete(node)
    return undefined
  }

  return walk(start)
}


export function relationFindings(
  decls: Map<string, RelDecl>, graph: Graph
): RelationFinding[] {
  const findings: RelationFinding[] = []

  const byRelation = new Map<string, Edge[]>()
  const pairs = new Set<string>()
  for (const e of graph.edges) {
    const list = byRelation.get(e.key)
    if (undefined === list) {
      byRelation.set(e.key, [e])
    }
    else {
      list.push(e)
    }
    pairs.add(e.key + ' ' + e.from + ' ' + e.to)
  }

  // Predicates in sorted order, so the findings arrive the same way
  // in both ports (the registry is insertion-ordered here, random in
  // Go).
  const names = [...decls.keys()].sort(cmpCodePoint)
  for (const name of names) {
    const decl = decls.get(name) as RelDecl
    const mine = byRelation.get(name) ?? []

    if (true === decl.acyclic) {
      const succ = new Map<string, string[]>()
      for (const e of mine) {
        const list = succ.get(e.from)
        const to = e.to
        if (undefined === list) {
          succ.set(e.from, [to])
        }
        else {
          list.push(to)
        }
      }
      for (const list of succ.values()) {
        list.sort(cmpCodePoint)
      }

      // The roots are visited in sorted order, and a node already
      // settled is not revisited, so one cycle is reported once and the
      // SAME one in both ports.
      const done = new Set<string>()
      const roots = [...succ.keys()].sort(cmpCodePoint)
      for (const from of roots) {
        const cycle = findCycle(from, succ, done)
        if (undefined !== cycle) {
          // The cycle's first node is a key of `succ`, and every key of
          // `succ` came from an edge's `from`, so the edge is there.
          const at = mine.find((e) => e.from === cycle[0]) as Edge
          findings.push({
            code: 'relation_cycle',
            relation: name,
            at: at.at,
            detail: cycle,
          })
          break
        }
      }
    }

    const inverses = [...decl.inverses].sort(cmpCodePoint)
    for (const inv of inverses) {
      for (const e of mine) {
        const to = e.to
        if (!pairs.has(inv + ' ' + to + ' ' + e.from)) {
          findings.push({
            code: 'relation_inverse_missing',
            relation: name,
            at: e.at,
            detail: [e.from, to, inv],
          })
        }
      }
    }
  }

  findings.sort((a, b) =>
    cmpCodePoint(a.at, b.at) || cmpCodePoint(a.code, b.code)
    || cmpCodePoint(a.detail.join(' '), b.detail.join(' ')))

  return findings
}


export function relationErrors(ctx: any, root: any): void {
  const decls: Map<string, RelDecl> = ctx._reldecls
  if (0 === decls.size) {
    return
  }
  const findings = relationFindings(decls, graphOf(root))
  for (const f of findings) {
    let node: any = root
    for (const seg of f.at.slice(2).split('.')) {
      // Graph atoms hold the field's value -- possibly nested, one
      // atom carrying another -- and the path steps through them
      // exactly as the graph walk does.
      while (true === node?.isGraphAtom) {
        node = node.held
      }
      node = node?.peg?.[seg]
    }
    ctx.adderr(makeNilErr(ctx, f.code, node, undefined, 'relate', {
      relation: f.relation,
      detail: f.detail.join(' -> '),
    }))
  }
}


// The relation checks for one document: evaluate, then report the
// same verdict generation enforces.
export function relationCheck(
  src: string, opts?: RelationOptions): RelationReport {
  const options = opts ?? {}
  const aontu = new Aontu(includeOpts(options))
  const ctx = aontu.ctx({ collect: true })
  const parseOpts = null == options.path ? undefined : { path: options.path }
  const root: any = aontu.unify(src, parseOpts, ctx)

  // A document that does not stand up is not a document with a bad
  // graph: the errors it already has are the answer, and blaming its
  // relations on top would be noise.
  if (0 < ctx.err.length || true === root?.isNil) {
    return {
      verdict: 'error',
      findings: [],
      errors: [failureFinding(ctx, options.path, root)],
    }
  }

  const decls: Map<string, RelDecl> = (ctx as any)._reldecls
  const counted = true === options.count ? { declared: decls.size } : {}
  if (0 === decls.size) {
    return { verdict: 'pass', findings: [], ...counted }
  }

  const gctx = ctx.clone({ err: [], collect: true })
  root.gen(gctx)
  if (0 < gctx.err.length) {
    return {
      verdict: 'error',
      findings: [],
      errors: [failureFinding(gctx, options.path, root)],
    }
  }

  const findings = relationFindings(decls, graphOf(root))
  return {
    verdict: 0 === findings.length ? 'pass' : 'fail',
    findings,
    ...counted,
  }
}
